ui/src/static/env.prod.js

```diff
@@ -0,0 +1,11 @@
+window.env = {
```

Comment:
Should these be commited to the codebase? They look like something that would rather be changed on infrastructure level

Reply:
I think they should but I could be wrong here. The idea was to allow people to modify these files (i.e., `build/static/env.js`) directly without having to `yarn build`. I think environment variables in the browser are not meant to contain secret values so there should probably be no worries in terms of security.

Reply:
It's just that it appears that you have to modify the code itself(make changes to these `env` files and commit them) in order to globally configure the application.... The idea of env variables seems to be the opposite of that - so that you could change the global configs at infrastructure level without the need to change the code when you deploy something somewhere.

Or am I misunderstanding how this is going to work?

Reply:
Right now if there is any problem with environment variables, I believe we need to recreate an image which makes the pod startup slow. That's because webpack injects environment variables at build time. What this PR does is define the environment variables at runtime so when something crashes because of misconfiguration for example, we don't need to build an new image. We can just change the `env.js` file directly and have the change be reflected right away. Once we are happy with the new config we can commit them. At least that's how I understand it.

> The idea of env variables seems to be the opposite of that - so that you could change the global configs at infrastructure level without the need to change the code when you deploy something somewhere.

Part 2 of this is to have a `config.yml` that defines these variables then copies them to `env.js` which if I understood correctly, will ease the interaction at the infrastructure level since that's how the rest of our services are configured. I may be wrong...

Reply:
Having spent a lot of time removing things referencing `taskcluster.net` from the codebase, I'm with @owlishDeveloper in being pretty skeptical of adding it back.  Especially since as of some date in the next few months, that domain won't work anymore.

> Once we are happy with the new config we can commit them.

We won't be able to check those in since they'll differ for every deployment.  There aren't even suitable default values that we can use for development, since there's no "main" deployment for which we could specify the TASKCLUSTER_ROOT_URL.

Maybe `yarn build` should not write an `env.js` file at all, and `yarn start` should be set up to write out a copy of the file based on the user's environment before running `neutrino start`, and failing if required fields like TASKCLUSTER_ROOT_URL aren't set.  Then part 2 is just doing a similar thing in the k8s startup process.

One tweak that might make that idea easier is to make all of those variables optional, or provide default values for them, so just running `yarn start` will get you something that works, even if it sometimes generates some bogus URLs (due to a default value for TASKCLUSTER_ROOT_URL).  Default values for the GraphQL endpoints could just be `/graphql` and `/subscription`, I think.  And the others seem to have decent default values already.

Reply:
> pretty skeptical of adding it back.

I'm not sure I follow. Adding back what?

> Maybe yarn build should not write an env.js file at all

Are you suggesting dropping the idea of `window.env`?

> yarn start should be set up to write out a copy of the file based on the user's environment before running neutrino start

Doesn't this patch already do that in .neutrinorc.js?

```js
// This code runs for both development and production
neutrino.use(copy, {
  patterns: [
    // ...
    {
      context: 'src/static',
      from: process.env.NODE_ENV === 'production' ? 'env.prod.js' : 'env.js',
      to: 'static/env.js'
    },
  ],
});
```

>  Default values for the GraphQL endpoints could just be /graphql and /subscription

Done.

Reply:
> > pretty skeptical of adding it back.
> 
> I'm not sure I follow. Adding back what?

Sorry -- adding files containing the string `taskcluster.net`.

> > Maybe yarn build should not write an env.js file at all
> 
> Are you suggesting dropping the idea of `window.env`?

No, I think `window.env` is the right idea -- I'm suggesting not *writing* `env.js` in `yarn build`, and instead always writing it before the server process is started.

> > yarn start should be set up to write out a copy of the file based on the user's environment before running neutrino start
> 
> Doesn't this patch already do that in .neutrinorc.js?

It does, but that's in `yarn build`.  I'm suggesting doing so only for `yarn start`, so the result of `yarn build` has no `env.js` at all.

I'm suggesting this to try to partition the steps of "building" and "running" the UI, even when developing.  So the output of the "building" part has no configuration, and the act of "running" the UI requires the user to supply some configuration.  I realize there's a little fuzziness there since `yarn start` also runs a build.

Reply:
> No, I think window.env is the right idea -- I'm suggesting not writing env.js in yarn build, and instead always writing it before the server process is started.

`yarn build` nukes `build/` and creates a new `build/` with fresh assets. A user who `rm -rf build/` would have a new `build/` directory generated with only one file `env.js` the next time they run `yarn start`. That looks a bit odd to me. Thoughts?

Reply:
Oh, I didn't realize that `yarn start` doesn't write to `build/`.  Does it just hold all that data in memory?  Is there a way to provide dynamic content for the HTTP resource `/static/env.js` during `yarn run` without writing it to disk as part of the build process?

Reply:
That's right. It holds all of that data in memory using [webpack-dev-server](https://github.com/webpack/webpack-dev-server) to support live reloading and fast access to assets.

> Is there a way to provide dynamic content for the HTTP resource /static/env.js during yarn run without writing it to disk as part of the build process?

I can't think of anything but maybe because I don't understand the problem. Did you have something in mind? Would [creating a webpack loader](https://webpack.js.org/contribute/writing-a-loader/) be helpful here?

Reply:
It might -- or maybe before calling `neutrino start`, write out an `env.js` that is in `.gitignore` so that it doesn't get committed.  Then `yarn build` can ensure that file's not present before calling `neutrino build`, so that the file isn't part of the build.